tools/halide_image.h

```diff
@@ -1,278 +1,673 @@
+    template<typename Array, size_t N>
+    int dimensionality_of_array(Array (&vals)[N]) {
+        return dimensionality_of_array(vals[0]) + 1;
+    }
+
+    int dimensionality_of_array(T &) {
+        return 0;
+    }
+
+    /** Check if any args in a parameter pack are zero */
+    template<typename ...Args>
+    bool any_zero(int first, Args... rest) {
+        if (first == 0) return true;
+        return any_zero(rest...);
+    }
+
+    bool any_zero() {
+        return false;
+    }
+
+public:
+    typedef T ElemType;
+
+    /** Public fields that halide_buffer_t has that buffer_t does not have. */
 
-        void dev_free() {
-            halide_device_free(NULL, &buf);
+    /** The dimensionality of the buffer */
+    int dimensions;
+
+    /** The halide type of the elements */
+    halide_type_t type;
+
+    /** halide_buffer_t accesses its shape via a pointer to an array
+     * of dimensions */
+    halide_dimension_t *dim = dim;
+
+    /** The total number of elements this buffer represents. Equal to
+     * the product of the extents */
+    size_t number_of_elements() const {
+        size_t s = 1;
+        for (int i = 0; i < dimensions; i++) {
+            s *= dim[i].extent;
         }
+        return s;
+    }
 
-        ~Contents() {
-            if (buf.dev) {
-                dev_free();
+    /** A pointer to the element with the lowest address. If all
+     * strides are positive, equal to the host pointer. */
+    uint8_t *begin() const {
+        ptrdiff_t index = 0;
+        for (int i = 0; i < dimensions; i++) {
+            if (dim[i].stride < 0) {
+                index += dim[i].stride * (dim[i].extent - 1);
             }
-            delete[] alloc;
         }
-    };
-
-    Contents *contents;
-
-    void initialize(int x, int y, int z, int w, bool interleaved) {
-        buffer_t buf = {0};
-        buf.extent[0] = x;
-        buf.extent[1] = y;
-        buf.extent[2] = z;
-        buf.extent[3] = w;
-        if (interleaved) {
-            buf.stride[0] = z;
-            buf.stride[1] = x*z;
-            buf.stride[2] = 1;
-            buf.stride[3] = x*y*z;
-        } else {
-            buf.stride[0] = 1;
-            buf.stride[1] = x;
-            buf.stride[2] = x*y;
-            buf.stride[3] = x*y*z;
+        return host + index * type.bytes();
+    }
+
+    /** A pointer to one beyond the element with the highest address. */
+    uint8_t *end() const {
+        ptrdiff_t index = 0;
+        for (int i = 0; i < dimensions; i++) {
+            if (dim[i].stride > 0) {
+                index += dim[i].stride * (dim[i].extent - 1);
+            }
+        }
+        index += 1;
+        return host + index * type.bytes();
+    }
+
+    /** The total number of bytes spanned by the data in memory. */
+    size_t size_in_bytes() const {
+        return (size_t)(end() - begin());
+    }
+
+    /** A pointer to the element at the given location. */
```

Comment:
I can also imagine that now that you mention it. People should use &im(x,
y, z) instead.

On Wed, Jul 27, 2016 at 4:43 PM, Steven Johnson <notifications@github.com>
wrote:

> In tools/halide_image.h
> <https://github.com/halide/Halide/pull/1402#discussion_r72544116>:
>
> > +        ptrdiff_t index = 0;
> > +        for (int i = 0; i < dimensions; i++) {
> > +            if (dim[i].stride > 0) {
> > +                index += dim[i].stride * (dim[i].extent - 1);
> > +            }
> > +        }
> > +        index += 1;
> > +        return host + index * type.bytes();
> > +    }
> > +
> > +    /** The total number of bytes spanned by the data in memory. */
> > +    size_t size_in_bytes() const {
> > +        return (size_t)(end() - begin());
> > +    }
> > +
> > +    /** A pointer to the element at the given location. */
>
> I will bet you a dollar we find people using this to get/set pixels. (Not
> saying it shouldn't exist. Just sayin'.)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/halide/Halide/pull/1402/files/c31b7ed6ad38c23104cb8006f2f532434ce250ff#r72544116>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAfdRqzXo4LP61FGsesn4qGPEXeyLkbWks5qZ-0XgaJpZM4JTMHs>
> .
>